onnxruntime/python/tools/symbolic_shape_infer.py

```diff
@@ -2829,8 +2829,99 @@ def infer_shapes(in_mp, int_max=2**31 - 1, auto_merge=False, guess_output_rank=F
         if not all_shapes_inferred:
             onnx.save_model(symbolic_shape_inference.out_mp_, "sym_shape_infer_temp.onnx", save_as_external_data=True)
             raise Exception("Incomplete symbolic shape inference")
+
+        SymbolicShapeInference.optimize_graph(symbolic_shape_inference)
```

Comment:
The utility is shared by many components. It shall only run shape inference, but not change graph.

You can move optimize_graph to another file like symbolic_shape_optimizer.py.

Reply:
thx for the reviewing, and good point!

this is just a draft to show perf gain in one 1p model, there are still some open questions to me: where to put the code, implement it in python or cpp, turn on by default or not(as shape inference may have error),...